Export GPU symbols for embedder

jwinarske opened the pull request:
This PR exports GPU symbols for the embedder library.

*List which issues are fixed by this PR. You must list at least one issue.*
https://github.com/flutter/flutter/issues/153196

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

bdero commented on shell/platform/embedder/embedder_exports.lst:
```diff
@@ -13,6 +13,7 @@
     kDartIsolateSnapshotInstructions;
     kDartVmSnapshotData;
     kDartVmSnapshotInstructions;
+    InternalFlutterGpu*;
```
```suggestion
    InternalFlutterGpu*;
    kInternalFlutterGpu*;
```

bdero commented:
Nice find.

bdero commented:
Are we properly roping the `flutter_gpu` package into the kernel?

Test-wise, I think adding something that forces one of the symbols to get called in the GLFW embedder smoketest would be suffice.

Perhaps just:
```diff
diff --git a/examples/glfw/main.dart b/examples/glfw/main.dart
index 557d020a1d..97c81efc6f 100644
--- a/examples/glfw/main.dart
+++ b/examples/glfw/main.dart
@@ -2,11 +2,15 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.

+import '../../lib/gpu/lib/gpu.dart' as gpu;
 import 'package:flutter/material.dart';
 import 'package:flutter/foundation.dart'
     show debugDefaultTargetPlatformOverride;

 void main() {
+  // Ensure Flutter GPU symbols are available by forcing the GPU context to instantiate.
+  gpu.gpuContext;
+
   // This is a hack to make Flutter think you are running on Google Fuchsia,
   // otherwise you will get an error about running from an unsupported platform.
   debugDefaultTargetPlatformOverride = TargetPlatform.fuchsia;
```

That's enough to trigger a symbol call.

bdero commented:
Ah actually, Impeller isn't enabled in that example, but we can still verify that the symbol is being called properly. Something like this:
```diff
diff --git a/examples/glfw/main.dart b/examples/glfw/main.dart
index 557d020a1d..ceb5eda137 100644
--- a/examples/glfw/main.dart
+++ b/examples/glfw/main.dart
@@ -2,11 +2,22 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.

+import 'package:flutter_gpu/gpu.dart' as gpu;
 import 'package:flutter/material.dart';
 import 'package:flutter/foundation.dart'
     show debugDefaultTargetPlatformOverride;

 void main() {
+  // Ensure Flutter GPU symbols are available by forcing the GPU context to instantiate.
+  try {
+    // ignore: unnecessary_statements
+    gpu.gpuContext; // Force the context to instantiate.
+  } catch (e) {
+    // If impeller is not enabled, make sure the exception isn't about symbols missing.
+    assert(e.toString().contains(
+        'Flutter GPU requires the Impeller rendering backend to be enabled.'));
+  }
+
   // This is a hack to make Flutter think you are running on Google Fuchsia,
   // otherwise you will get an error about running from an unsupported platform.
   debugDefaultTargetPlatformOverride = TargetPlatform.fuchsia;
```

jwinarske commented:
@bdero makes sense

jwinarske commented:
The relative path is not working.  It should be relative to the myapp/lib folder.  I tried all variations, no dice.

Moved to `import 'package:flutter_gpu/gpu.dart' as gpu;` and added `flutter pub add flutter_gpu --sdk=flutter` to run.sh.

That part is fine, only the app hits an unrelated problem:
```
In order to run your application, type:

  $ cd myapp
  $ flutter run

Your application code is in myapp/lib/main.dart.

Resolving dependencies... 
Downloading packages... 
+ flutter_gpu 0.0.0 from sdk flutter
  material_color_utilities 0.11.1 (0.12.0 available)
Changed 1 dependency!
1 package has newer versions incompatible with dependency constraints.
Try `flutter pub outdated` for more information.
Resolving dependencies... 
Downloading packages... 
  material_color_utilities 0.11.1 (0.12.0 available)
Got dependencies!
1 package has newer versions incompatible with dependency constraints.
Try `flutter pub outdated` for more information.

lib/main.dart:1:1: Error: The specified language version is too high. The highest supported language version is 3.5.
// Copyright 2013 The Flutter Authors. All rights reserved.
^
../../../../../../../../flutter/packages/flutter/lib/src/semantics/semantics.dart:2925:7: Error: No named parameter with the name 'linkUrl'.
      linkUrl: data.linkUrl?.toString() ?? '',
      ^^^^^^^
Target kernel_snapshot_program failed: Exception
```

jwinarske commented:
```
$ flutter doctor -v
[✓] Flutter (Channel main, 3.25.0-1.0.pre.63, on Ubuntu 20.04.6 LTS 5.15.0-118-generic, locale en_US.UTF-8)
    • Flutter version 3.25.0-1.0.pre.63 on channel main at /home/tcna/workspace-automation/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fbed99fd4b (86 minutes ago), 2024-08-21 05:25:21 -0700
    • Engine revision b94e009703
    • Dart version 3.6.0 (build 3.6.0-164.0.dev)
    • DevTools version 2.39.0-dev.15

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.30.2
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] VS Code (version 1.92.2)
    • VS Code at /snap/code/current/usr/share/code
    • Flutter extension version 3.94.0

[✓] Connected device (2 available)
    • Linux (desktop)            • linux              • linux-x64 • Ubuntu 20.04.6 LTS 5.15.0-118-generic
    • Toyota homescreen (mobile) • desktop-homescreen • linux-x64 • homescreen x86_64

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

jwinarske commented:
```
$ flutter pub outdated
Showing outdated packages.
[*] indicates versions that are not the latest available.

Package Name              Current  Upgradable  Resolvable  Latest  

direct dependencies: all up-to-date.

dev_dependencies: all up-to-date.

transitive dependencies: 
material_color_utilities  *0.11.1  *0.11.1     *0.11.1     0.12.0  
all dependencies are up-to-date.
```

jwinarske commented:
@bdero 
I just wrapped my due diligence; problem was on my end.  We are good to go.  Assert hits if symbols are not available.  This is the happy path output:
```
Creating project myapp...
Resolving dependencies in `myapp`... 
Downloading packages... 
Got dependencies in `myapp`.
Wrote 18 files.

All done!
You can find general documentation for Flutter at: https://docs.flutter.dev/
Detailed API documentation is available at: https://api.flutter.dev/
If you prefer video documentation, consider: https://www.youtube.com/c/flutterdev

In order to run your application, type:

  $ cd myapp
  $ flutter run

Your application code is in myapp/lib/main.dart.

Resolving dependencies... 
Downloading packages... 
+ flutter_gpu 0.0.0 from sdk flutter
  material_color_utilities 0.11.1 (0.12.0 available)
Changed 1 dependency!
1 package has newer versions incompatible with dependency constraints.
Try `flutter pub outdated` for more information.
Resolving dependencies... 
Downloading packages... 
  material_color_utilities 0.11.1 (0.12.0 available)
Got dependencies!
1 package has newer versions incompatible with dependency constraints.
Try `flutter pub outdated` for more information.

/home/tcna/workspace-automation/app/engine/src/flutter/examples/glfw/debug
[ERROR:flutter/shell/platform/embedder/embedder_surface_gl_skia.cc(123)] Could not create a resource context for async texture uploads. Expect degraded performance. Set a valid make_resource_current callback on FlutterOpenGLRendererConfig.
[ERROR:flutter/shell/platform/embedder/embedder_surface_gl_skia.cc(123)] Could not create a resource context for async texture uploads. Expect degraded performance. Set a valid make_resource_current callback on FlutterOpenGLRendererConfig.
flutter: The Dart VM service is listening on http://127.0.0.1:34919/E_SicMZ-fEE=/
```

bdero reviewed (approved):
LGTM
